File: sample/java/12_sort_visualizer.java
final class _12_sort_visualizer {
    private _12_sort_visualizer() {
    }


    // 12: Sample that outputs intermediate states of bubble sort as a GIF.

    public static java.util.ArrayList<Long> render(java.util.ArrayList<Long> values, long w, long h) {
        java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(w * h);
        long n = ((long)(values.size()));
        double bar_w = ((double)(w)) / ((double)(n));
        for (long i = 0L; i < n; i += 1L) {
            long x0 = PyRuntime.__pytra_int(((double)(i)) * bar_w);
            long x1 = PyRuntime.__pytra_int((((double)(i + 1L))) * bar_w);
            if (((x1) <= (x0))) {
                x1 = x0 + 1L;
            }
            long bh = PyRuntime.__pytra_int(((double)(((Long)(values.get((int)((((i) < 0L) ? (((long)(values.size())) + (i)) : (i)))))))) / ((double)(n)) * ((double)(h)));
            long y = h - bh;
            for (y = y; y < h; y += 1L) {
                for (long x = x0; x < x1; x += 1L) {
                    frame.set((int)((((y * w + x) < 0L) ? (((long)(frame.size())) + (y * w + x)) : (y * w + x))), 255L);
                }
            }
        }
        return PyRuntime.__pytra_bytearray(frame);
    }

    public static void run_12_sort_visualizer() {
        long w = 320L;
        long h = 180L;
        long n = 124L;
        String out_path = "sample/out/12_sort_visualizer.gif";
        double start = time.perf_counter();
        java.util.ArrayList<Long> values = new java.util.ArrayList<Long>();
        long i = 0L;
        for (i = 0L; i < n; i += 1L) {
            values.add((i * 37L + 19L) % n);
        }
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>(java.util.Arrays.asList(render(values, w, h)));
        long frame_stride = 16L;
        long op = 0L;
        for (i = 0L; i < n; i += 1L) {
            boolean swapped = false;
            for (long j = 0L; j < n - i - 1L; j += 1L) {
                if (((((Long)(values.get((int)((((j) < 0L) ? (((long)(values.size())) + (j)) : (j))))))) > (((Long)(values.get((int)((((j + 1L) < 0L) ? (((long)(values.size())) + (j + 1L)) : (j + 1L))))))))) {
                    long __swap_tmp_0 = ((Long)(values.get((int)((((j) < 0L) ? (((long)(values.size())) + (j)) : (j))))));
                    values.set((int)((((j) < 0L) ? (((long)(values.size())) + (j)) : (j))), ((Long)(values.get((int)((((j + 1L) < 0L) ? (((long)(values.size())) + (j + 1L)) : (j + 1L)))))));
                    values.set((int)((((j + 1L) < 0L) ? (((long)(values.size())) + (j + 1L)) : (j + 1L))), __swap_tmp_0);
                    swapped = true;
                }
                if (((op % frame_stride) == (0L))) {
                    frames.add(render(values, w, h));
                }
                op += 1L;
            }
            if ((!swapped)) {
                break;
            }
        }
        gif.save_gif(out_path, w, h, frames, gif.grayscale_palette(), 3L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(((long)(frames.size()))));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/julia/12_sort_visualizer.jl
include(joinpath(@__DIR__, "built_in", "py_runtime.jl"))

include(joinpath(@__DIR__, "std", "time.jl"))
include(joinpath(@__DIR__, "utils", "gif.jl"))

# 12: Sample that outputs intermediate states of bubble sort as a GIF.

function render(values, w, h)
    frame = __pytra_bytearray((w * h))
    n = length(values)
    bar_w = (w / n)
    for i in 0:n - 1
        x0 = __pytra_int((i * bar_w))
        x1 = __pytra_int(((i + 1) * bar_w))
        if (x1 <= x0)
            x1 = (x0 + 1)
        end
        bh = __pytra_int(((values[__pytra_idx(i, length(values))] / n) * h))
        y = (h - bh)
        for y in y:h - 1
            for x in x0:x1 - 1
                frame[__pytra_idx(((y * w) + x), length(frame))] = 255
            end
        end
    end
    return __pytra_bytes(frame)
end

function run_12_sort_visualizer()
    w = 320
    h = 180
    n = 124
    out_path = "sample/out/12_sort_visualizer.gif"
    
    start = perf_counter()
    values = Any[]
    i = nothing
    for i in 0:n - 1
        push!(values, (((i * 37) + 19) % n))
    end
    frames = [render(values, w, h)]
    frame_stride = 16
    
    op = 0
    for i in 0:n - 1
        swapped = false
        for j in 0:(((n - i) - 1)) - 1
            if (values[__pytra_idx(j, length(values))] > values[__pytra_idx((j + 1), length(values))])
                __swap_tmp_0 = values[__pytra_idx(j, length(values))]
                values[__pytra_idx(j, length(values))] = values[__pytra_idx((j + 1), length(values))]
                values[__pytra_idx((j + 1), length(values))] = __swap_tmp_0
                swapped = true
            end
            if ((op % frame_stride) == 0)
                push!(frames, render(values, w, h))
            end
            op = op + 1
        end
        if (!(swapped))
            break
        end
    end
    save_gif(out_path, w, h, frames, grayscale_palette(), 3, 0)
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", length(frames))
    __pytra_print("elapsed_sec:", elapsed)
end


run_12_sort_visualizer();
